File: lab4/get_addrs.pl
#!/usr/bin/perl

use warnings;
use strict;
my $file = "/etc/dhcp/dhcpd.conf";
my $regex_ip = '^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$';
my $regex_mac = '^([0-9A-F]{2}[:]){5}([0-9A-F]{2})$';
my @host =();
my @mac =();
my @ip =();

if (@ARGV != 0)
{
	die "Usage: sudo ./get_addr";
}
else
{
	open(my $IN, "<$file") or die "couldn't open $file";
	while(my $line = <$IN>)
	{
		
		chomp($line);
		&f_host($line);
		&f_macadd($line);
		&f_ipadd($line);
	}
    #check size of the array
    if ($#host != $#mac or $#mac != $#ip)
    {
        die "ERROR: Check Array size"
    }
    else
    {
        #output
    	for (my $i=0; $i<=($#host); $i++ )
    	{
    		print "$host[$i] $mac[$i] $ip[$i] \n";
    	}
    }   
}


#find hostname
sub f_host{
	if($_[0] =~ /host/i)
	{
		#print $_[0];
		my @words = split(/host/i, $_[0]);
		@words = split(/{/, $words[1] );
		&chopper($words[0]);
	#	print "$words[0] ";
		push(@host, $words[0]);
		return @host;
	}
	

}
#find MAC address
sub f_macadd{
	if($_[0] =~ /hardware ethernet/i)
	{
	#	print "$_[0]\n";
		my @words = split(/hardware\s+ethernet/i, $_[0]);
        #       print "slashed: $words[1]\n";
		@words = split(/;/, $words[1] );
		&chopper($words[0]);
		#check mac reguex
		if ($words[0] =~ /$regex_mac/i)
		{
			#print "$words[0] ";
			push(@mac, $words[0]);
			return @mac;
		}
		else
		{
			print "Not Valid MAC address\n";
		}
	}
}
#find IP address
sub f_ipadd{
	if($_[0] =~ /fixed-address/i)
	{
		my @words = split(/fixed-address/i, $_[0]);
		@words = split(/;/, $words[1] );
		&chopper($words[0]);
	        
		if ($words[0] =~ /$regex_ip/)
                {
                        #print "$words[0] ";
			push(@ip, $words[0]);
			return @ip;
                }
                else
                {
                        print "Not Valid IP address\n";
                }
	}
}

#take out all space
sub chopper{
	        $_[0] =~ s/^\s*//;
                $_[0] =~ s/\s*$//;
	}

File: lab4/get_addrs.py
#!/usr/bin/python
import sys, re, subprocess, os

file= '/etc/dhcp/dhcpd.conf'
host=[]
mac=[]
ip=[]

def f_host(x):
    if re.search("host", x):
        print "original:", x
        x = re.split(ur"\s*host\s*", x)
        x = re.split(ur"\s*{", x[1])
        print x[0]
        #host.append(x[0])
def f_mac(x):
    if re.search("hardware ethernet", x):
        x = re.split(ur"hardware\s+ethernet", x)
        x = chopper(x)
        #print "MAC",x[0], "[end]"
        if re.search(r"^([0-9A-F]{2}[:]){5}([0-9A-F]{2})$", x[0]):
            #print x[0]
            mac.append(x[0])
        else:
            print "Error: not a valid MAC address"

def f_ip(x):
    if re.search("fixed-address", x):
        x = re.split(ur"fixed-address", x)
        x = chopper(x)
        #print "IP:", x[0],"[end]"
        if re.search(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$", x[0]):
            #print x[0]  
            ip.append(x[0])
        else:
            print "Error not a vaid IP address"

def chopper(x):
    x = re.split(";", x[1])
    x = re.split(r"\s*", x[0])
    x = re.split(r"\s*$", x[1])
    #print x[0]
    return x

if len(sys.argv) != 1:
    print "Usage: sudo ./get_addr.py"
else:
    fchk = os.path.exists("/etc/dhcp/dhcpd.conf")
    
    if fchk:
        pass
    else:
        "Error: dhcpd.conf does not exist."
        sys.exit()
    try:
        file = open("/etc/dhcp/dhcpd.conf", "r")
        line = file.readline()
    except IOError:
        print "Could'nt open a file"
        sys.exit()
    while line:
        line=line.rstrip("\n")
        f_host(line)
        f_mac(line)
        f_ip(line)
        
        line = file.readline()
 
    #output
    if len(host) != len(mac) or len(mac) != len(ip):
        print "Error: list size does not match "
        sys.exit()
    else:
        for i in range(len(host)):
            print host[i], mac[i], ip[i]

 
